packaging/metadata.py

```diff
@@ -0,0 +1,416 @@
+            urls = {}
+            for url in cast(Iterable[str], self.project_url):
+                key, _, value = url.partition(",")
+                urls[key.strip()] = value.strip()
+        _setattr(self, "project_url", ImmutableMapping(urls))
+
+        # Dataclasses don't enforce data types at runtime
+        if not isinstance(self.requires_python, SpecifierSet):
+            requires_python = self._parse_requires_python(self.requires_python)
+            _setattr(self, "requires_python", requires_python)
+        if self.version and not isinstance(self.version, Version):
+            _setattr(self, "version", Version(self.version))
+
+        if self.dynamic:
+            values = (_normalize_field_name_for_dynamic(f) for f in self.dynamic)
+            dynamic = frozenset(v for v in values if self._validate_dynamic(v))
+            _setattr(self, "dynamic", dynamic)
+
+    @property
+    def metadata_version(self) -> str:
+        """
+        The data structure is always compatible with the latest approved
+        version of the spec, even when parsing files that use previous versions.
+        """
+        return "2.2"
+
+    @classmethod
+    @lru_cache(maxsize=None)
+    def _fields(cls) -> Collection[str]:
+        return frozenset(f.name for f in dataclasses.fields(cls))
+
+    @classmethod
+    def _read_pkg_info(cls, pkg_info: bytes) -> Dict[str, Any]:
+        """Parse PKG-INFO data."""
+
+        msg = message_from_bytes(pkg_info, EmailMessage, policy=cls._PARSING_POLICY)
+        info = cast(EmailMessage, msg)
+
+        attrs: Dict[str, Any] = {}
+        for key in info.keys():
+            field = key.lower().replace("-", "_")
+            if field in cls._UPDATES:
+                field = cls._UPDATES[field]
+
+            value = str(info.get(key))  # email.header.Header.__str__ handles encoding
+
+            if field == "keywords":
+                attrs[field] = " ".join(value.splitlines()).split(",")
+            elif field == "description":
+                attrs[field] = cls._unescape_description(value)
+            elif field.endswith("email"):
+                attrs[field] = cls._parse_emails(value)
+            elif field in cls._MULTIPLE_USE:
+                attrs[field] = (str(v) for v in info.get_all(key))
+            elif field in cls._fields():
+                attrs[field] = value
+
+        if "description" not in attrs:
+            attrs["description"] = info.get_content(content_manager=raw_data_manager)
+
+        return attrs
+
+    @classmethod
+    def from_pkg_info(cls: Type[T], pkg_info: bytes) -> T:
+        """Parse PKG-INFO data."""
+
+        return cls(**cls._read_pkg_info(pkg_info))
+
+    @classmethod
+    def from_dist_info_metadata(cls: Type[T], metadata_source: bytes) -> T:
+        """Parse METADATA data."""
+
+        attrs = cls._read_pkg_info(metadata_source)
+        _attrs = attrs.copy()
+        _attrs.pop("description", None)
+
+        if attrs.get("dynamic"):
```

Comment:
```suggestion
        if "dynamic" in attrs:
```

Otherwise your key access below could fail.

Reply:
Thank you very much for the suggestion, I think this is a good improvement since in this situation we don't expect the `attrs` to contain an empty/None value for "dynamic"

I don't understand your comment though `Otherwise your key access below could fail`... My understanding is that 

```python
if attrs.get("dynamic"):
```

is roughly equivalent to:

```python
if "dynamic" in attrs and attrs["dynamic"]
```

is there any chance that `attrs.get("dynamic")` will return a Truth-y value without `"dynamic"` being an existing key in `attrs`?